gpu4pyscf/grad/tdrks_ris.py

```diff
@@ -0,0 +1,365 @@
+    if with_k:
+        vj0, vk0 = mf.get_jk(mol, dmzoo, hermi=0)
+        vj1 = mf_J.get_j(mol, dmxpy + dmxpy.T, hermi=0)
+        vk1 = mf_K.get_k(mol, dmxpy + dmxpy.T, hermi=0)
+        vk2 = mf_K.get_k(mol, dmxmy - dmxmy.T, hermi=0)
+        if not isinstance(vj0, cp.ndarray):
+            vj0 = cp.asarray(vj0)
+        if not isinstance(vk0, cp.ndarray):
+            vk0 = cp.asarray(vk0)
+        if not isinstance(vj1, cp.ndarray):
+            vj1 = cp.asarray(vj1)
+        if not isinstance(vk1, cp.ndarray):
+            vk1 = cp.asarray(vk1)
+        if not isinstance(vk2, cp.ndarray):
+            vk2 = cp.asarray(vk2)
+        vj = cp.stack((vj0, vj1))
+        vk = cp.stack((vk0, vk1, vk2))
+        vk *= hyb
+        if omega != 0:
+            vk0 = mf.get_k(mol, dmzoo, hermi=0, omega=omega)
+            vk1 = mf_K.get_k(mol, dmxpy + dmxpy.T, hermi=0, omega=omega)
+            vk2 = mf_K.get_k(mol, dmxmy - dmxmy.T, hermi=0, omega=omega)
+            if not isinstance(vk0, cp.ndarray):
+                vk0 = cp.asarray(vk0)
+            if not isinstance(vk1, cp.ndarray):
+                vk1 = cp.asarray(vk1)
+            if not isinstance(vk2, cp.ndarray):
+                vk2 = cp.asarray(vk2)
+            vk += cp.stack((vk0, vk1, vk2)) * (alpha - hyb)
+        veff0doo = vj[0] * 2 - vk[0] + f1oo[0]
+        veff0doo += td_grad.solvent_response(dmzoo)
+        wvo = reduce(cp.dot, (orbv.T, veff0doo, orbo)) * 2
+        if singlet:
+            veff = vj[1] * 2 - vk[1]
+        else:
+            veff = - vk[1]
+        veff += td_grad.solvent_response(dmxpy + dmxpy.T)
+        veff0mop = reduce(cp.dot, (mo_coeff.T, veff, mo_coeff))
+        wvo -= contract("ki,ai->ak", veff0mop[:nocc, :nocc], xpy) * 2
+        wvo += contract("ac,ai->ci", veff0mop[nocc:, nocc:], xpy) * 2
+        veff = -vk[2]
+        veff0mom = reduce(cp.dot, (mo_coeff.T, veff, mo_coeff))
+        wvo -= contract("ki,ai->ak", veff0mom[:nocc, :nocc], xmy) * 2
+        wvo += contract("ac,ai->ci", veff0mom[nocc:, nocc:], xmy) * 2
+    else:
+        vj0 = mf.get_j(mol, dmzoo, hermi=1)
+        vj1 = mf_J.get_j(mol, dmxpy + dmxpy.T, hermi=1)
+        if not isinstance(vj0, cp.ndarray):
+            vj0 = cp.asarray(vj0)
+        if not isinstance(vj1, cp.ndarray):
+            vj1 = cp.asarray(vj1)
+        vj = cp.stack((vj0, vj1))
+
+        veff0doo = vj[0] * 2 + f1oo[0] 
+        wvo = reduce(cp.dot, (orbv.T, veff0doo, orbo)) * 2
+        if singlet:
+            veff = vj[1] * 2
+        else:
+            veff = 0
+        veff0mop = reduce(cp.dot, (mo_coeff.T, veff, mo_coeff))
+        wvo -= contract("ki,ai->ak", veff0mop[:nocc, :nocc], xpy) * 2
+        wvo += contract("ac,ai->ci", veff0mop[nocc:, nocc:], xpy) * 2
+        veff0mom = cp.zeros((nmo, nmo))
+
+    # set singlet=None, generate function for CPHF type response kernel
+    # TODO: LR-PCM TDDFT
+    vresp = td_grad.base._scf.gen_response(singlet=None, hermi=1)
```

Comment:
* The new API td_grad.base.gen_response is available.

* If hasattr(td_grad.base, 'with_solvent'): raise